CMakeLists.txt

```diff
@@ -548,6 +548,24 @@ else()
   endif()
   list(APPEND b2
     variant=release link=static threading=multi cxxflags=${BOOST_CFLAGS})
+  if(NOT CMAKE_HOST_SYSTEM_PROCESSOR STREQUAL CMAKE_SYSTEM_PROCESSOR)
+    # we are crosscompiling
+    if(CMAKE_CXX_COMPILER_ID STREQUAL GNU)
+      set(b2_cc gcc)
+    elseif(CMAKE_CXX_COMPILER_ID STREQUAL Clang)
+      set(b2_cc clang)
+    else()
+      message(SEND_ERROR "unknown compiler: ${CMAKE_CXX_COMPILER_ID}")
+    endif()
+    # edit the config.jam so, b2 will be able to use the specified toolset
+    execute_process(
+      COMMAND
+      sed -i
+      "s|using ${b2_cc} ;|using ${b2_cc} : ${CMAKE_SYSTEM_PROCESSOR} : ${CMAKE_CXX_COMPILER} ;|"
+      ${PROJECT_SOURCE_DIR}/src/boost/project-config.jam)
```

Comment:
@cbodley our boost building process is not friendly to the use case you mentioned above: it creates b2/bjam, settings files and other asserts in **source directory**. for example, the executable of "b2", "bjam", and the header directory of "boost" are all created right in the source tree of `boost`. please check its `.gitignore` for more details. in other words, the cmake building process(es) will interfere with each other.

maybe i could find a way to fix this so we can have a true out-of-source build. but maybe not today. as ExternalProject's building process is run on demand, and we can hardly ready boost's header directories without building it. and cmake does check for the existence of header directories at seeing `include_directories()`. so we might need to run cmake in a tricky way to address this issue, for example, by instructing cmake to launch itself just for building boost. but again, we probably should address this in another PR.